### PR TITLE
Cs/sc 3069 subscribe to ds changes

### PR DESCRIPTION
**Pull request by Charl1996**
Opening as draft since I'm busy checking how we can better manage migrations in this repo.

The changes introduced in this PR is as follows:
1. Add and invoke task to subscribe to a specific datasource on HQ
2. Adds a new API for handling OAuth with superset as the authorization server.
3. Add decorator (`require_oauth`) to check access token on `/hq_webhook/change/` endpoint.

### CODE REVIEW

File: hq_superset/models.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+from superset import db
+from authlib.integrations.sqla_oauth2 import OAuth2ClientMixin
+from werkzeug.security import check_password_hash, generate_password_hash
+
+
+@dataclass
+class DataSetChange:
+    action: Literal['upsert', 'delete']
+    data_source_id: str
+    data: dict[str, Any]
```
Comment (SmittieC) [inline hq_superset/models.py:15]: I wonder if we might not want to have another dataclass for this `data` field. That way it will do its own validation and we will know exactly what keys we expect

File: hq_superset/utils.py [hunk 2]
```diff
@@ -190,23 +197,236 @@ def get_datasource_file(path):
...
+
+    try:
+        array_values = ast.literal_eval(string_array)
+    except ValueError:
+        return []
+
+    if isinstance(array_values, tuple):
+        array_values = list(array_values)
+
+    # Test for corner cases
```
Comment (SmittieC) [inline hq_superset/utils.py:261]: is this being tested?

**Comment by Charl1996** [comment]
I hope I rebased correctly. It seems like it.

File: hq_superset/views.py [hunk 3]
```diff
@@ -247,8 +232,8 @@ def post_dataset_change(self) -> FlaskResponse:
 
         try:
             request_json = json.loads(request.get_data(as_text=True))
-            change = DataSetChange(**request_json)  # raises TypeError
-            update_dataset(change)  # raises ValueError
+            change = DataSetChange(**request_json)
+            update_dataset(change)
```
Comment (Charl1996) [inline hq_superset/views.py:236]: @kaapstorm Should this be commented out?

File: .idea/.gitignore
[the remark this answers is not in the document]
Comment (Charl1996) [reply .idea/.gitignore:1]: Oh, I forgot to remove that (I did see it and forgot to remove it). Thanks for the suggestion about gitignore_global, I'll do that!

[the remark this answers is not in the document]
Comment (Charl1996) [reply .idea/.gitignore:1]: Removed

File: hq_superset/views.py (hunk 3 above)
Comment (kaapstorm) [reply hq_superset/views.py:236]: No? That's the call that does the work.

File: hq_superset/api.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+def query_client(client_id):
+    return HQClient.get_by_client_id(client_id)
+
+
+def save_token(token, request):
+    client = request.client
+    client.revoke_tokens()
+
+    expires_at = datetime.utcnow() + timedelta(seconds=ONE_DAY_SECONDS)
```
Comment (kaapstorm) [inline hq_superset/api.py:27]:
```suggestion
    expires_at = datetime.utcnow() + timedelta(days=1)
```

File: hq_superset/models.py [hunk 5]
```diff
@@ -11,3 +17,71 @@ class DataSetChange:
...
+        return client.domain
+
+    def is_expired(self):
+        return datetime.utcnow() > self.expires_at
+
+    def is_revoked(self):
+        return self.revoked
+
+    def get_scope(self):
+        return self.scope
```
Comment (kaapstorm) [inline hq_superset/models.py:87]: It's not obvious to me why you created a method that returns a property that is already public, and it doesn't look like a subclass will override this method. Do we need it?

**Review by kaapstorm — APPROVED** [review]
All looking good. Just a few minor comments.

File: hq_superset/models.py (hunk 5 above)
Comment (Charl1996) [reply hq_superset/models.py:87]: The `get_scope` is a method that the `require_oauth` decorator is looking for on the Token class, so it has to be defined as such. I'll add a comment and make that clear.

File: hq_superset/models.py [hunk 6]
```diff
@@ -11,3 +17,71 @@ class DataSetChange:
...
+    def get_by_domain(cls, domain):
+        return db.session.query(HQClient).filter_by(domain=domain).first()
+
+    @classmethod
+    def get_by_client_id(cls, client_id):
+        return db.session.query(HQClient).filter_by(client_id=client_id).first()
+
+    @classmethod
+    def create_domain_client(cls, domain: str):
+        client_secret = str(uuid.uuid4())
```
[the remark this answers is not in the document]
Comment (Charl1996) [reply hq_superset/models.py:49]: Ah, thanks for the suggestion! Turns out uniqueness != cryptographically secure!
